pallets/asset-manager/src/migrations.rs

```diff
@@ -0,0 +1,416 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonbeam.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::{AssetIdType, AssetTypeId, AssetTypeUnitsPerSecond, Config};
+use frame_support::{
+	pallet_prelude::PhantomData,
+	storage::migration::{remove_storage_prefix, storage_key_iter},
+	traits::{Get, OnRuntimeUpgrade},
+	weights::Weight,
+	Blake2_128Concat,
+};
+use sp_std::convert::TryInto;
+//TODO sometimes this is unused, sometimes its necessary
+use sp_std::vec::Vec;
+use xcm::latest::prelude::*;
+
+pub struct UnitsWithAssetType<T>(PhantomData<T>);
+impl<T: Config> OnRuntimeUpgrade for UnitsWithAssetType<T> {
+	fn on_runtime_upgrade() -> Weight {
+		log::info!(target: "UnitsWithAssetType", "actually running it");
+		let pallet_prefix: &[u8] = b"AssetManager";
+		let storage_item_prefix: &[u8] = b"AssetIdUnitsPerSecond";
+
+		// Read all the data into memory.
+		// https://crates.parity.io/frame_support/storage/migration/fn.storage_key_iter.html
+		let stored_data: Vec<_> = storage_key_iter::<T::AssetId, u128, Blake2_128Concat>(
+			pallet_prefix,
+			storage_item_prefix,
+		)
+		.collect();
+
+		let migrated_count: Weight = stored_data
+			.len()
+			.try_into()
+			.expect("There are between 0 and 2**64 mappings stored.");
+
+		log::info!(target: "UnitsWithAssetType", "Migrating {:?} elements", migrated_count);
+
+		// Now remove the old storage
+		// https://crates.parity.io/frame_support/storage/migration/fn.remove_storage_prefix.html
+		remove_storage_prefix(pallet_prefix, storage_item_prefix, &[]);
+
+		// Assert that old storage is empty
+		assert!(storage_key_iter::<T::AssetId, u128, Blake2_128Concat>(
+			pallet_prefix,
+			storage_item_prefix
+		)
+		.next()
+		.is_none());
+
+		// Write to the new storage with removed and added fields
+		for (asset_id, units) in stored_data {
+			// Read the assetType for the assetId
+			if let Some(asset_type) = AssetIdType::<T>::get(&asset_id) {
+				// Populate with assetType as key
+				AssetTypeUnitsPerSecond::<T>::insert(&asset_type, units)
+			}
+		}
+
+		log::info!(target: "UnitsWithAssetType", "almost done");
+
+		// Return the weight used. For each migrated mapping there is a read to get it into
+		// memory, a read to get assetType and
+		// a write to clear the old stored value, and a write to re-store it.
+		let db_weights = T::DbWeight::get();
+		migrated_count.saturating_mul(2 * db_weights.write + 2 * db_weights.read)
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn pre_upgrade() -> Result<(), &'static str> {
+		use frame_support::traits::OnRuntimeUpgradeHelpersExt;
+
+		let pallet_prefix: &[u8] = b"AssetManager";
+		let storage_item_prefix: &[u8] = b"AssetIdUnitsPerSecond";
+
+		// We want to test that:
+		// There are no entries in the new storage beforehand
+		// The same number of mappings exist before and after
```

Comment:
Strictly speaking, this wasn't enforced previously (see my "many to one" vs "one to one" comments above). That shouldn't be a problem though, because the current on-chain storage has no such mappings, right?